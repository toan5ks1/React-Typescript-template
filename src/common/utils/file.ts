import accepts from "attr-accept";
import { KeyMediaTypes } from "src/common/constants/const";
import { MIMETypes } from "src/common/constants/enum";

import { MediaType } from "manabie-bob/enum_pb";

import { FILE_SIZE_UNITS, GIGABYTE, MEGABYTE } from "../constants/file-size";

import isNil from "lodash/isNil";
import type { Media } from "src/__generated__/root-types";

export interface ValidateTypeSchema {
    minSize?: number;
    maxSize: number;
    accept: string | string[];
}

export interface ConvertByteReturn {
    size: number;
    unit: string;
}

export const getFileExtension = (filename: string) => {
    return filename.split(".").pop() || "";
};

export const getFileNameWithoutExtension = (filename: string) => {
    return filename.replace(`.${getFileExtension(filename)}`, "");
};

export function getFileSize(e: File) {
    return e.size;
}

//convert byte to MB or GB
export function convertByte(byte?: number): ConvertByteReturn {
    if (!byte) return { size: 0, unit: "" };

    if (byte >= GIGABYTE) {
        return { size: Math.ceil(byte / GIGABYTE), unit: FILE_SIZE_UNITS.GB };
    } else {
        return { size: Math.ceil(byte / MEGABYTE), unit: FILE_SIZE_UNITS.MB };
    }
}

export const toBase64 = async (file: File) => {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        reader.onload = () => resolve(reader.result);
        reader.onerror = (error) => reject(error);
        reader.readAsDataURL(file);
    });
};
const base64ToPayload = (base64: string): string => {
    return base64.split(",")[1];
};

export const fileToPayload = async (file: File) => {
    const data = await toBase64(file);
    return base64ToPayload(String(data));
};

export const fileMatchSize = (file: File, minSize?: number, maxSize?: number): boolean => {
    if (!isNil(file.size)) {
        if (minSize && maxSize) return file.size >= minSize && file.size <= maxSize;
        if (minSize) return file.size >= minSize;
        if (maxSize) return file.size <= maxSize;
    }
    return true;
};

export const fileAccepted = (file: File, accept: string | string[]): boolean => {
    return file.type === "application/x-moz-file" || accepts(file, accept);
};

export const filterFiles = (files: File[], validateSchema: ValidateTypeSchema[]): File[] => {
    const schemaSize = validateSchema.length;
    let acceptedFiles: File[] = [];
    if (!files || !files.length) return acceptedFiles;

    files.forEach((file) => {
        for (let i = 0; i < schemaSize; i++) {
            const { minSize, maxSize, accept } = validateSchema[i];
            const isAccepted = fileMatchSize(file, minSize, maxSize) && fileAccepted(file, accept);
            if (isAccepted) acceptedFiles.push(file);
        }
    });
    return acceptedFiles;
};

export type FilterUploadFiles = File | Media;

export const isMedia = (media: FilterUploadFiles): media is Media => {
    return (media as Media).media_id !== undefined;
};

export const filterUploadFiles = (files: FilterUploadFiles[]) => {
    const alreadyUploadedMediaIds: string[] = [];
    const filesNeedToBeUploaded: File[] = [];
    for (let i = 0; i < files.length; i++) {
        const current = files[i];
        if (isMedia(current)) {
            alreadyUploadedMediaIds.push(current.media_id);
            continue;
        }
        if (current instanceof File) {
            filesNeedToBeUploaded.push(current);
        }
    }
    return { alreadyUploadedMediaIds, filesNeedToBeUploaded };
};

export const MapMediaMIMETypes = {
    [MIMETypes.PDF]: KeyMediaTypes?.MEDIA_TYPE_PDF,
    [MIMETypes.VIDEO]: KeyMediaTypes?.MEDIA_TYPE_VIDEO,
    [MIMETypes.AUDIO]: KeyMediaTypes?.MEDIA_TYPE_AUDIO,
    [MIMETypes.IMAGE]: KeyMediaTypes?.MEDIA_TYPE_IMAGE,
};

export function getCheckFileTypeFn(fileType: MIMETypes) {
    return ({ type }: { type?: string }) => {
        return accepts({ type }, fileType) || MapMediaMIMETypes[fileType] === type;
    };
}

export const isImage = getCheckFileTypeFn(MIMETypes.IMAGE);
export const isVideo = getCheckFileTypeFn(MIMETypes.VIDEO);
export const isPdf = getCheckFileTypeFn(MIMETypes.PDF);
export const isAudio = getCheckFileTypeFn(MIMETypes.AUDIO);

export const groupBy = ({ files }: { files: File[] }) => {
    let others: File[] = [];
    const videos = files.filter((file) => {
        if (isVideo({ type: file.type })) return file;
        others.push(file);
    });
    return { others, videos };
};
export const findMediaType = (type: string): MediaType => {
    if (isImage({ type })) return MediaType.MEDIA_TYPE_IMAGE;
    if (isPdf({ type })) return MediaType.MEDIA_TYPE_PDF;
    if (isAudio({ type })) return MediaType.MEDIA_TYPE_AUDIO;
    if (isVideo({ type })) return MediaType.MEDIA_TYPE_VIDEO;
    return MediaType.MEDIA_TYPE_NONE;
};

export interface AttachmentInfo {
    type: MediaType;
    name: string;
    resource: string;
}
export const matchingUpsertMedia = ({
    url,
    type,
    name,
}: {
    url: string;
    type: string;
    name: string;
}): AttachmentInfo => ({
    name,
    type: findMediaType(type),
    resource: url,
});
