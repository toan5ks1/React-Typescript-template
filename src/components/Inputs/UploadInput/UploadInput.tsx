import { PropsWithChildren, ReactNode } from "react";

import { useDropzone, DropzoneOptions } from "react-dropzone";

import { Box } from "@mui/material";
import FormUploadFilePlaceholder from "src/components/Forms/FormUploadFile/FormUploadFilePlaceholder";
import TypographyBase from "src/components/Typographys/TypographyBase";
import { useFormProps } from "src/providers/FormPropsProvider";

import useDropAccepted, { UseDropAcceptedProps } from "src/hooks/useDropAccepted";
import useUploadFiles from "src/hooks/useUploadFiles";

export interface UploadInputProps extends PropsWithChildren<DropzoneOptions>, UseDropAcceptedProps {
    multiple?: boolean;
    uploading?: boolean;
    description?: ReactNode;
    rootDefaultPadding?: boolean;
    descDefaultMargin?: boolean;
}

//todo: allow upload multiple files
const UploadInput = (props: UploadInputProps) => {
    const {
        accept,
        maxSize,
        minSize,
        multiple = false,
        uploading,
        description,
        onChange,
        customFilterFn,
        disabled,
        rootDefaultPadding = true,
        descDefaultMargin = true,
    } = props;

    const { onDropAccepted } = useDropAccepted({
        customFilterFn,
        onChange,
    });
    const { readOnly } = useFormProps();

    const { onDropRejected } = useUploadFiles(maxSize);

    const { getRootProps, getInputProps, isDragActive } = useDropzone({
        accept,
        maxSize,
        minSize,
        multiple,
        onDropAccepted,
        onDropRejected,
        disabled: uploading || readOnly || disabled,
    });

    return (
        <>
            <Box
                {...getRootProps()}
                sx={(theme) => ({
                    padding: rootDefaultPadding ? theme.spacing(3) : theme.spacing(2, 3),
                    ...{
                        width: "100%",
                        minHeight: 82,
                        border: `1px dashed rgba(33, 150, 243, 0.5)`,
                        borderRadius: theme.spacing(0.5),
                        background: "rgba(33, 150, 243, 0.08)",
                        display: "flex",
                        justifyContent: "center",
                        alignItems: "center",
                        cursor: "pointer",
                        "&:focus": {
                            outline: "unset",
                        },
                    },
                })}
                data-testid="UploadInput"
            >
                <input data-testid="UploadInput__inputFile" {...getInputProps()} />
                <FormUploadFilePlaceholder
                    isDragActive={isDragActive}
                    uploading={uploading || false}
                />
            </Box>
            {description && (
                <TypographyBase
                    sx={(theme) => ({
                        display: "inline-block",
                        color: theme.palette.text.secondary,
                        margin: descDefaultMargin
                            ? theme.spacing(1.5, 0)
                            : theme.spacing(1.1, 0, 0, 0),
                    })}
                    variant="caption"
                    data-testid="UploadInput__description"
                >
                    {description}
                </TypographyBase>
            )}
        </>
    );
};

export default UploadInput;
