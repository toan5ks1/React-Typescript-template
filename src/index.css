body {
    margin: 0;
    font-family: "Roboto", sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    overflow-x: hidden;
}

code {
    font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace;
}

body,
html,
#root {
    margin: 0;
    padding: 0;
    width: 100%;
    min-height: 100vh;
    background-color: #fff;
}

#root > div:first-child {
    margin: 0;
    padding: 0;
    width: 100%;
    min-height: 100vh;
    box-sizing: border-box;
}

.list-page > div > div:last-child {
    overflow: auto;
}

table td {
    max-width: 300px;
    overflow-wrap: break-word;
}

/* Override css background input autocomplete - Designer wants it */
input:-webkit-autofill,
input:-webkit-autofill:hover,
input:-webkit-autofill:focus,
textarea:-webkit-autofill,
textarea:-webkit-autofill:hover,
textarea:-webkit-autofill:focus,
select:-webkit-autofill,
select:-webkit-autofill:hover,
select:-webkit-autofill:focus {
    -webkit-box-shadow: 0 0 0px 1000px #fff inset;
    transition: background-color 5000s ease-in-out 0s;
}

/* styled not affectted, so we use this, check src/components/BrightcoveVideo/BrightcoveVideo.tsx */
.bc-player-default_default {
    width: 100%;
    height: 100%;
}
