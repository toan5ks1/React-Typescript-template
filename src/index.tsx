import "src/styles/mui-classname-setup";

import { StrictMode } from "react";

import { enableMapSet } from "immer";
import ReactDOM from "react-dom";
import enableSidebarMethod from "src/squads/architecture/internals/sidebar";
import enableManaEventEmitter from "src/squads/root/src/internals/mana-event-emitter";

import App from "./app";
import AppProvider from "./app/AppProvider";
import "./index.css";
import { enableReport } from "./internals/stackdriver";
import { enableLogger } from "./internals/warner";

enableMapSet();
enableReport();
enableLogger();

enableSidebarMethod();
enableManaEventEmitter();

ReactDOM.render(
    <StrictMode>
        <AppProvider>
            <App />
        </AppProvider>
    </StrictMode>,
    document.getElementById("root")
);
