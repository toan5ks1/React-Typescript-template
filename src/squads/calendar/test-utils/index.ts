export { default as TestThemeProvider } from "./TestThemeProvider";
export { default as TestApp } from "./TestApp";
export { default as AppProvider } from "./AppProvider";
