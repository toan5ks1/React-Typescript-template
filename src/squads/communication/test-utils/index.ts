export { default as TestThemeProvider } from "./TestThemeProvider";
export { default as TestApp } from "./TestApp";
