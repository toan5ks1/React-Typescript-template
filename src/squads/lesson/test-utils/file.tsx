interface CreateFile {
    (fileName?: string, fileProperty?: FilePropertyBag): File;
}

export const createFile: CreateFile = (
    fileName = "FILE_NAME",
    fileProperty = {
        type: "image/png",
    }
) => {
    return new File([new ArrayBuffer(1)], fileName, fileProperty);
};
