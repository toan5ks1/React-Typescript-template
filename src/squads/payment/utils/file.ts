import { parse } from "papaparse";
import { arrayHasItem } from "src/common/utils/other";

export async function parseCSVFile<T>(file: File): Promise<T[]> {
    return new Promise((resolve, reject) => {
        parse(file, {
            header: true,
            transform: (fielValue) => {
                return fielValue.trim();
            },
            complete: (result) => {
                if (arrayHasItem(result.errors)) {
                    reject(result.errors);
                    return;
                }

                resolve(result.data as T[]);
            },
            error: (err) => {
                reject(err);
            },
        });
    });
}
