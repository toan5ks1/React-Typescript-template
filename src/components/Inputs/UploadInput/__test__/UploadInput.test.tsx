import { TestThemeProvider } from "src/test-utils";
import { createFile } from "src/test-utils/file";
import { TestQueryWrapper } from "src/test-utils/react-hooks";

import UploadInput, { UploadInputProps } from "../UploadInput";

import { render, fireEvent, waitFor } from "@testing-library/react";

const inputTestId: string = "UploadInput__inputFile";

const descriptionTestId: string = "UploadInput__description";

const renderUtil = (props: UploadInputProps) => {
    return render(
        <TestQueryWrapper>
            <TestThemeProvider>
                <UploadInput {...props} />
            </TestThemeProvider>
        </TestQueryWrapper>
    );
};

jest.mock("src/internals/configuration/configuration");
jest.mock("src/hooks/useShowSnackbar", () => ({
    __esModule: true,
    default: () => jest.fn(),
}));

describe("<UploadInput /> UI", () => {
    it("should match snapshot", async () => {
        const props: UploadInputProps = {
            accept: "image/png",
            description: "DESCRIPTION",
        };

        const { container } = renderUtil(props);

        expect(container).toMatchSnapshot();
    });

    it("should render correct UI", () => {
        const props: UploadInputProps = {
            description: "DESCRIPTION",
        };

        const { getByTestId } = renderUtil(props);

        expect(getByTestId(descriptionTestId).textContent).toMatch(/DESCRIPTION/gi);
    });
});

describe("<UploadInput /> handler", () => {
    it("should call exactly the callback", async () => {
        const props: UploadInputProps = {
            onChange: jest.fn(),
            accept: "image/png",
        };

        const { getByTestId, container } = renderUtil(props);

        expect(container).toMatchSnapshot();

        const inputTarget = getByTestId(inputTestId) as HTMLInputElement;
        const file = createFile();
        Object.defineProperty(inputTarget, "files", { value: [file] });

        fireEvent.change(inputTarget);

        await waitFor(() => {
            expect(props.onChange).toBeCalled();
        });
        expect(props.onChange).toHaveBeenCalledWith([file]);
    });
});
