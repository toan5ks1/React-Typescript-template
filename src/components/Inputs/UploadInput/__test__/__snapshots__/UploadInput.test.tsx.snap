// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadInput /> UI should match snapshot 1`] = `
.emotion-0 {
  padding: 24px;
  width: 100%;
  min-height: 82px;
  border: 1px dashed rgba(33, 150, 243, 0.5);
  border-radius: 4px;
  background: rgba(33, 150, 243, 0.08);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-0:focus {
  outline: unset;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  margin-right: 16px;
  color: #90CAF9;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  font-size: 3.125rem;
}

.emotion-4 {
  margin: 0;
  font-size: 0.875rem;
  font-weight: 400;
  letter-spacing: 0.15px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.43;
  color: #757575;
}

.emotion-5 {
  color: #42A5F5;
}

.emotion-6 {
  margin: 0;
  font-size: 0.75rem;
  font-weight: 400;
  letter-spacing: 0.4px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.66;
  color: #757575;
  display: inline-block;
  margin: 12px 0px;
}

<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="UploadInput"
    role="button"
    tabindex="0"
  >
    <input
      accept="image/png"
      autocomplete="off"
      data-testid="UploadInput__inputFile"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="MuiBox-root emotion-1"
    >
      <div
        class="MuiBox-root emotion-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="CloudUploadOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM19 18H6c-2.21 0-4-1.79-4-4 0-2.05 1.53-3.76 3.56-3.97l1.07-.11.5-.95C8.08 7.14 9.94 6 12 6c2.62 0 4.88 1.86 5.39 4.43l.3 1.5 1.53.11c1.56.1 2.78 1.41 2.78 2.96 0 1.65-1.35 3-3 3zM8 13h2.55v3h2.9v-3H16l-4-4z"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 emotion-4"
      >
        resources.input.dragAndDropFileHere
         
        <span
          class="MuiBox-root emotion-5"
        >
          resources.input.browse
        </span>
      </p>
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-caption emotion-6"
    data-testid="UploadInput__description"
  >
    DESCRIPTION
  </span>
</div>
`;

exports[`<UploadInput /> handler should call exactly the callback 1`] = `
.emotion-0 {
  padding: 24px;
  width: 100%;
  min-height: 82px;
  border: 1px dashed rgba(33, 150, 243, 0.5);
  border-radius: 4px;
  background: rgba(33, 150, 243, 0.08);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-0:focus {
  outline: unset;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  margin-right: 16px;
  color: #90CAF9;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  font-size: 3.125rem;
}

.emotion-4 {
  margin: 0;
  font-size: 0.875rem;
  font-weight: 400;
  letter-spacing: 0.15px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.43;
  color: #757575;
}

.emotion-5 {
  color: #42A5F5;
}

<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="UploadInput"
    role="button"
    tabindex="0"
  >
    <input
      accept="image/png"
      autocomplete="off"
      data-testid="UploadInput__inputFile"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="MuiBox-root emotion-1"
    >
      <div
        class="MuiBox-root emotion-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="CloudUploadOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM19 18H6c-2.21 0-4-1.79-4-4 0-2.05 1.53-3.76 3.56-3.97l1.07-.11.5-.95C8.08 7.14 9.94 6 12 6c2.62 0 4.88 1.86 5.39 4.43l.3 1.5 1.53.11c1.56.1 2.78 1.41 2.78 2.96 0 1.65-1.35 3-3 3zM8 13h2.55v3h2.9v-3H16l-4-4z"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 emotion-4"
      >
        resources.input.dragAndDropFileHere
         
        <span
          class="MuiBox-root emotion-5"
        >
          resources.input.browse
        </span>
      </p>
    </div>
  </div>
</div>
`;
